json-api/dev1

tribbloid opened the pull request:
JSON file compatibility

contains the following 3 patches:
- core
- cargo update and JetBrains RustRover setup
- public direct load_runtime_data API

on 1 extra test case `single_job_with_single_task_json` is added

tribbloid committed:
first commit, can read json file

tribbloid committed:
cargo update

git ignore rustrover dir

tribbloid committed:
2 functions for load_runtime_data are now public

jaredgorski commented:
Code looks fine to me on shallow glance; I think we'd want test(s) specifically on `tests/profile.rs` to more explicitly cover this sort of change. Anyway, I have one question and one statement:

**Question:**
What's the use case here? Since JSON doesn't allow real line breaks, it's not at all ergonomic for writing profiles by hand. Are you attempting to machine-generate a profile and have some limitation where you can't convert to YAML prior to passing to arpx?

**Statement:**
Per [the philosophy](https://github.com/jaredgorski/arpx?tab=readme-ov-file#note-on-the-philosophy-behind-this-project), specifically "Arpx doesn't seek to be ... general-purpose", I'm a bit skeptical that expanding input config formats to JSON (TOML, XML, .ini ...) fulfills a core need, especially with how easy it is to convert between JSON and YAML these days ([`mikefarah/yq` does it OOTB](https://mikefarah.gitbook.io/yq/usage/convert)). I'm not saying I can't be convinced, just that I like keeping things almost painfully slim. Worst case, your more permissive fork becomes the superstar crate mine never was 😸

tribbloid commented:
From our side (https://github.com/hpvdt/HMD-Air), it is slightly faster to orchestrate through a foreign function interface (FFI) than a file, in fact, arpx was chosen precisely for it's interoperability on the DLL level.

JSON interface for FFI is often easier with less line breaks and platform-dependent specifications

jaredgorski commented:
Ah, that makes sense. FFI is probably the only use case that might convince me to change my mind.

Just to push the issue... don't you think it's to your advantage to depend entirely on your fork, rather than my primary repo? You could publish your forked crate to https://crates.io/ and now you've got your own controlled, [vendored](https://macwright.com/2021/03/11/vendor-by-default) version of arpx that you can modify to your needs, without worry that somebody with access to my project will break the code under your feet.

If you indeed prefer to put this on mainline arpx, I'm not opposed. I'm _alright_ with using the file extension, but I think an improvement would actually be to just silently failover from one to the other... that way we're not depending on the file name to tell us the format to expect. Does that make sense?

My requests then, if you're set on adding this to mainline arpx:
- see if you can make the deserialize code failover from YAML to JSON rather than using the file extension to denote the format
- tests on `test/profile.rs` (the module being changed), rather than using a test on `runtime.rs` (not the module being changed).

I'll also mention that I haven't touched the release process in a few years, so there's probably some rust (ha) to sand off before we get the CI/CD working smoothly.

Let me know your thoughts.

tribbloid committed:
the only json test has been moved to profile test suite

load_runtime for file with unknown extension will try to parse yaml first then parse json

tribbloid commented:
just submitted a new patch that does both:

- moved the only test case to be under profile test suite
- for file with unknown extension (not one of yaml/yml/json), it will try to parse it as yaml then failover to json
  - this can be extended to all files, but stacking 2 parsing errors may break a lot of test condition, is it really what you want?
  
  I wasn't aware of a canonical solution to report parsing error with multiple causes at the same time, maybe I should figure that part out
